### ISSUE TITLE
Bug: MemberExpression.object should be Expression, not LeftHandSideExpression

### ISSUE THREAD

**Issue by chris-grabcad** [typescript-eslint/typescript-eslint#6192]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Repro Code

```typescript
function isExpressionShould(node: TSESTree.MemberExpression): boolean {
    const object = node.object;
    const property = node.property;

    return (
        object?.type === AST_NODE_TYPES.BinaryExpression &&
        isIdentifier(property) &&
        property.name === "should"
    );
}
```

### ESLint Config

```javascript
N/A
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expect my code to compile without errors

### Actual Result

 error TS2367: This condition will always return 'false' since the types 'AST_NODE_TYPES.ArrayExpression | AST_NODE_TYPES.ArrayPattern | AST_NODE_TYPES.ArrowFunctionExpression | AST_NODE_TYPES.CallExpression | AST_NODE_TYPES.ClassExpression | AST_NODE_TYPES.FunctionExpression | ... 15 more ... | AST_NODE_TYPES.TSTypeAssertion' and 'AST_NODE_TYPES.BinaryExpression' have no overlap.

57         object?.type === AST_NODE_TYPES.BinaryExpression &&
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.

### Additional Info

I am working on a repo which uses node 14.16.0 and cannto upgrade to node 14.17.0 (yet) because it is used in an electron 12 project.

I'm trying to write a lint rule to catch sillinesses like `should(a === b).be.true()` etc, and another way it can be written is `(a === b).should.be.true()`, however I cannot get this check to compile which is a problem (I'll probably have to patch the types in my repo to get it working for now)

I checked this code with ast explorer

```ts
(apple === banana).should.be.true();
```

and I get this
![image](https://user-images.githubusercontent.com/5262051/206760188-f310cae5-edeb-493a-8182-d0b121e22acd.png)

I've also caught it in a vscode debugger with the values

<img width="880" alt="image" src="https://user-images.githubusercontent.com/5262051/206760990-cbfe31e3-4495-457d-8077-fff938c02ced.png">

<img width="464" alt="image" src="https://user-images.githubusercontent.com/5262051/206761270-bb100477-b733-4043-b851-e689e186aba7.png">




### Versions

<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `4.33.0` |
| `@typescript-eslint/parser`            | `4.31.2` |
| `@typescript-eslint/scope-manager`     | `4.33.0` |
| `@typescript-eslint/typescript-estree` | 4.31.2` |
| `@typescript-eslint/type-utils`        | N/A ??? |
| `@typescript-eslint/visitor-keys`        | `4.31.2` |
| `@typescript-eslint/types`             | `4.31.2` |
| `TypeScript`                           | `4.2.3` |
| `ESLint`                               | `7.26.0` |
| `node`                                 | `14.16.0` |

**Comment by bradzacher** [comment]
@JoshuaKGoldberg I wouldn't consider a bugfix a breaking change here beacuse:
1) it's just a type fix, not a runtime change
2) the types don't match the runtime behaviour, meaning consumers could already be broken
